core/shared/src/main/scala/zio/ConfigProvider.scala

```diff
@@ -35,19 +35,26 @@ trait ConfigProvider { self =>
    * single configuration value.
    */
   final def nested(name: String): ConfigProvider =
-    new ConfigProvider {
-      def load[A](config: Config[A])(implicit trace: Trace): IO[Config.Error, A] = self.load(config.nested(name))
-    }
+    ConfigProvider.fromFlat(self.flatten.nested(name))
 
   /**
    * Returns a new config provider that preferentially loads configuration data
    * from this one, but which will fall back to the specified alterate provider
    * if there are any issues loading the configuration from this provider.
    */
   final def orElse(that: ConfigProvider): ConfigProvider =
-    new ConfigProvider {
-      def load[A](config: Config[A])(implicit trace: Trace): IO[Config.Error, A] =
-        self.load(config).orElse(that.load(config))
+    ConfigProvider.fromFlat(self.flatten.merge(that.flatten))
```

Comment:
Are we sure we want to change the semantic of `orElse` versus just introducing `merge`? In some cases you may want to say, "Load this thing entirely, if you happen to randomly load some part of it, that's not valid, ignore it and try this other source instead." That's quite different than `merge`.

Reply:
It is a good question.

My thought was that applying the "or else" logic to the values in the `ConfigProvider` was significantly more intuitive. This issue originally arose because in doing something like `Config.string("key").optional` and `envProvider.orElse(propsProvider)` it was counterintuitive that this would return `None` if there was no value in `envProvider` but there was in `propsProvider`. 

That was potentially due in part to `orElse` currently being the only binary operator to compose `ConfigProvider` values but I think this does still seem potentially surprising. Related to this, I thought there was value in having one way of doing things unless we thought both were really valuable.

One further issue is that I'm not sure the old `orElse` semantics make sense in the world in which every `ConfigProvider` has an underlying `Flat` implementation. `Flat` splits enumerating the keys and loading a value into separate operators. But the old `orElse` semantics imply that we need to load all the values in the left provider and make sure they can be loaded successfully before knowing whether we should enumerate the keys in the left provider or the keys in the right provider.

Reply:
All right, I think I'm open to that. The final question is: should we even introduce `merge` since it's another name for `orElse`?

Reply:
I think that makes sense, especially since this way of "merging" is left biased whereas normal merging of key value pairs would normally be right biased if you had to pick one.